server/src/main/java/org/elasticsearch/cluster/coordination/NodeJoinExecutor.java

```diff
@@ -138,7 +140,7 @@ public ClusterState execute(BatchExecutionContext<JoinTask> batchExecutionContex
         DiscoveryNodes.Builder nodesBuilder = DiscoveryNodes.builder(newState.nodes());
         Map<String, CompatibilityVersions> compatibilityVersionsMap = new HashMap<>(newState.compatibilityVersions());
         Map<String, Set<String>> nodeFeatures = new HashMap<>(newState.nodeFeatures());
-        Set<String> allNodesFeatures = ClusterFeatures.calculateAllNodeFeatures(nodeFeatures.values());
+        Set<String> allAssumedNodesFeatures = calculateAllAssumedClusterFeatures(newState.nodes(), nodeFeatures);
```

Comment:
This variable name seems a bit scrambled. They're assumed features, not assumed nodes. Isn't this `allNodesAssumedFeatures`?

Reply:
Wondering if we can improve naming here... this isn't just assumed features of all nodes, but all explicit features + the assumed ones. So it's the effectiveClusterFeatures as seen by this node, while nodes of the next major will see something different.

Reply:
I like the `effective` naming. I'll use that.